build.rs

```diff
@@ -103,6 +103,13 @@ mod asm {
         if let Arch::Arm(arch) = arch {
             define(Define::bool("ARCH_ARM", arch == ArchArm::Arm32));
             define(Define::bool("ARCH_AARCH64", arch == ArchArm::Arm64));
+
+            if arch == ArchArm::Arm64 {
+                define(Define::bool("HAVE_DOTPROD", features.contains("dotprod")));
```

Comment:
Ideally it should be based on what can be compiled. If I set `HAVE_I8MM` to true, I get build errors such as these on an M2-based Mac:
```
 cargo:warning=<instantiation>:292:9: error: instruction requires: i8mm

  cargo:warning=        usdot v3.4s, v26.16b, v7.4b[1]
```

Reply:
We want to compile with all possible assembly for a given target and then do runtime feature detection.  What exactly is error on `i8mm` there; what tools is reporting it?  It should be able to compile on whatever platform/CPU as long as it's not trying to execute `i8mm` instructions on a CPU without `i8mm`.  So I don't how it's becoming a build error for you.

Reply:
Sorry, the previous error snapshot was a bit too short. The error occurs when processing assembly files:
```
  running: env -u IPHONEOS_DEPLOYMENT_TARGET "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-arch" "arm64" "-I" "." "-I" "/Users/bossen/Development/isrg/rav1d/target/release/build/rav1d-a41c145a1115d7a9/out" "-Wall" "-Wextra" "-o" "/Users/bossen/Development/isrg/rav1d/target/release/build/rav1d-a41c145a1115d7a9/out/src/arm/64/mc_dotprod.o" "-c" "src/arm/64/mc_dotprod.S"
```

Reply:
> The error occurs when processing assembly files

Hmm.  Is this some sort of cross-compilation failure?  `cc` should be able to cross-compile to a CPU that has `i8mm` without itself being a CPU that has `i8mm`, right?

Reply:
For the meson/C build, 2cf2558 adds testing `cc` with a whole bunch of options to determine what works and what doesn't. So I wouldn't expect every toolchain to support `i8mm`.

Reply:
We should not put toolchain version requirements that are simply based on performance.

That said, won't the `CARGO_CFG_TARGET_FEATURE` contain the supported features for the target? So if we're cross-compiling this will still enable CPU features for the target, not the host. So what's the problem with this approach?

Reply:
The target CPU architecture should specify which extensions are allowed for ARM. We should not be adding runtime detection that upstream is not doing (and they don't seem to be doing any here).

Reply:
Wait, we do runtime checks on cpu flags. What is the problem here? This LGTM

Reply:
If we compile on a CPU without these features and then run on a CPU with these features, we don't get to use the features to run faster since they weren't compiled.

Reply:
Ok, @kkysen has (very patiently) helped me understand the issue here, maybe I can rephrase that here in different words.

Rust `target-features` defaults to a minimal set of features _required_ by the target triple. We want to build with more ASM than that and check features at runtime (which this PR includes). However, we can't rely on target features to determine what instructions the compiler is aware of and supports. I think we might need a tiny compile-time toolchain feature support check in build.rs for this (which is what I think Khyber was saying above).

I however _do_ think detection is the best way to do this, we don't want to just not build on toolchains that don't support i8mm (i.e. the OS X compiler?).

Reply:
I finally figured out how to build the `i8mm` code on MacOS. And it actually executes on my machine!